### PR TITLE
Ieee802154 changes for better & future hw features support

### PR DESCRIPTION
**Pull request by tbursztyka**

### CODE REVIEW

**Comment by galak** [comment]
While not related, I added #4099 and make a note here since you're doing some ieee802154 work :)

File: subsys/net/ip/l2/ieee802154/ieee802154.c [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "ieee802154_frame.h"
```
Comment (jukkar) [inline subsys/net/ip/l2/ieee802154/ieee802154.c]: The commit message is not proper English, could you change to something like:

```
net/ieee802154: Apply hw filters when applicable

If the hw supports filtering, then apply the ieee address, short address or PAN ID filters.
```

File: include/net/ieee802154_radio.h [hunk 2]
```diff
@@ -71,15 +71,6 @@ struct ieee802154_radio_api {
 			  enum ieee802154_filter_type type,
```
Comment (jukkar) [inline include/net/ieee802154_radio.h]: This could also be merged with "drivers/ieee802154: Stop exposing old radio API functions" as this commit looks a bit redundant now. In that case the commit subject could be changed "drivers/ieee802154: Remove old radio API functions".

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good, minor nit picks.

File: include/net/ieee802154_radio.h (hunk 2 above)
Comment (tbursztyka) [reply include/net/ieee802154_radio.h]: I won't merge, as the other patch touches drivers and this one the public API in include. I like to keep things per-domain as much as possible (it's not always possible but here it is.).

File: subsys/net/ip/l2/ieee802154/ieee802154.c (hunk 1 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154.c]: ok

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+		if (radio->set_filter(iface->dev,
+				      IEEE802154_FILTER_TYPE_SHORT_ADDR,
+				      &filter) != 0) {
+			NET_WARN("Could not apply short address filter");
+		}
+	}
+}
+
+static inline void ieee802154_filter_pan_id(struct net_if *iface,
```
Comment (dleach02) [inline subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Why is this needed when there was already a radio->set_pan_id()? Are we planning on abstracting all fot he radio functions behind an ieee802154_<func>? Because it looks a bit strange to see code mixing the two when they ultimately touch a radio function.

**Review by dleach02 — APPROVED** [review]

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 3 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: It disappears in further patch. I did this order of patches for proper API deprecation

File: include/net/net_pkt.h [hunk 4]
```diff
@@ -116,6 +116,7 @@ struct net_pkt {
 
 #if defined(CONFIG_NET_L2_IEEE802154)
 	u8_t ieee802154_rssi;
+	u8_t ieee802154_lqi;
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/net/net_pkt.h]: It's not pedantic, it's actually required yes

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 3 above)
Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Is the intent then to add ieee802154_<func>() interface routines for all of the function points defined in ieee802154_radio_api structure? I would recommend doing this because it is a bit confusing to have a mix of access paths.

File: include/net/net_pkt.h [hunk 5]
```diff
@@ -123,6 +123,7 @@ struct net_pkt {
...
```
Comment (dleach02) [inline include/net/net_pkt.h]: There is a comment block at the beginning of this structure to remind authors that if you change this structure you need to also change the net_pkt_clone() function in net_pkt.c.

File: subsys/net/ip/l2/ieee802154/ieee802154.c [hunk 6]
```diff
@@ -233,7 +234,8 @@ static enum net_verdict ieee802154_recv(struct net_if *iface,
 	}
 
 	if (mpdu.mhr.fs->fc.frame_type == IEEE802154_FRAME_TYPE_BEACON) {
-		return ieee802154_handle_beacon(iface, &mpdu);
+		return ieee802154_handle_beacon(iface, &mpdu,
+						net_pkt_ieee802154_lqi(pkt));
```
Comment (dleach02) [inline subsys/net/ip/l2/ieee802154/ieee802154.c]: This would be cleaner if we just passed the pkt to handle_beacon instead of the mpdu and let ieee802154_handle_beacon() extract the lqi from the net_pkt directly.

File: include/net/ieee802154_radio.h [hunk 7]
```diff
@@ -24,6 +24,14 @@ extern "C" {
  * @{
  */
 
+enum ieee802154_hw_caps {
+	IEEE802154_HW_FCS		= BIT(0),
+	IEEE802154_HW_PROMISC		= BIT(1),
+	IEEE802154_HW_FILTER		= BIT(2),
+	IEEE802154_HW_CSMA		= BIT(3),
```
Comment (jfischer-phytec-iot) [inline include/net/ieee802154_radio.h]: Do you also want to insert IEEE802154_HW_ACK_REQ for the device with the capability to perform TX->RX(ACK) sequences?

File: include/net/ieee802154_radio.h [hunk 8]
```diff
@@ -24,6 +24,14 @@ extern "C" {
...
+	IEEE802154_HW_2_4_GHZ		= BIT(4),
```
Comment (jfischer-phytec-iot) [inline include/net/ieee802154_radio.h]: IEEE802154_HW_UWB?

Comment (jfischer-phytec-iot) [inline include/net/ieee802154_radio.h]: What do you think about:
```
IEEE802154_HW_ BPSK = BIT(4),
IEEE802154_HW_O-QPSK =BIT(5),
IEEE802154_HW_2450_DSSS = BIT(6),
IEEE802154_HW_UWB_LB = BIT(7), /* USB low band */
```
?

**Review by jfischer-no — APPROVED** [review]
Looks good, a few remarks about hw_caps

File: subsys/net/ip/l2/ieee802154/ieee802154.c (hunk 6 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154.c]: Yep, will do it in another patch though

File: include/net/net_pkt.h (hunk 5 above)
Comment (tbursztyka) [reply include/net/net_pkt.h]: Cloning is interesting for TCP/IP layer. I don't think anybody would need to keep track of rssi/lqi (which are valid only on rx path) beyond L2.

File: include/net/ieee802154_radio.h (hunk 7 above)
Comment (tbursztyka) [reply include/net/ieee802154_radio.h]: Yes, this will be necessary, as this PR https://github.com/zephyrproject-rtos/zephyr/pull/4115 reminded me as well.

File: include/net/ieee802154_radio.h (hunk 8 above)
Comment (tbursztyka) [reply include/net/ieee802154_radio.h]: Hum, not sure I'll add those. I went for the most generic one first, those affecting directly the L2 layer and the drivers we already have.

About the *PSK, DSSS and UWB (Ultra-Wide Band btw, it has nothing to do with USB ;) ), I am not convinced L2 will ever touch things about modulation. For instance, about cc1200, switching from one to the other requires a lot of parts, fully hw centric, no really possible to make generic. So basically, it is then up to the user to provide the RF settings (a default one and more if he wants to) and select the one he wants, feeding the driver with it.

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 3 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: No, it's just fine as it is right now. A structure which has already xyz_ as prefix  does not need to repeat that prefix on its attribute.

File: subsys/net/ip/l2/ieee802154/ieee802154.c (hunk 6 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154.c]: Actually I am revising the idea of changing this, it is better as it is right now where handle_beacon gets the pre-parsed mpdu and the lqi as parames instead of getting the net_pkt where it would require to re-parse all over again the 15.4 frame.

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 3 above)
Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: I don't think you understood my question. What I'm wondering is if the long term intent is to have management functions that abstract the interface to all of the function points defined in ieee802154_radio_api structure?

File: subsys/net/ip/l2/ieee802154/ieee802154.c (hunk 6 above)
Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154.c]: your perogative. Just seems cleaner to pass the net_pkt instead of the mpdu which would then carry the packet values.

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 3 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Indeed I misunderstood.

Hum about management functions I guess you are not talking about net_mgmt. (There would be no use to expose the radio API elsewhere than in the L2)

Do you mean having function ieee802154_foo(iface, ...params...) { iface->dev->radio_api->foo(...params...); } ?

If so, it really depends on the use cases. I did that for filtering because it was easier to have those instead of repeating most of the code here and there. It really depends.

Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Yes. That is what I was initially referring to because I find it a bit confusing to see two different methods for interacting with the radio; these ieee802154_foo() functions and the radio_api->foo functions. There was at least one place in the code where the two methods are next to each other.

I think that standardizing on one method is good as it will help future folks grep the code logic. It is also what I thought you were setting out to do with this initial set of changes.

Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: > There was at least one place in the code where the two methods are next to each other.

Accessing the same radio function?

This patch was only about filters. But why not generalizing for the others.
It's nicer to read ieee802154_set_channel() than radio->set_channel() yes. Will see in another patch.

File: include/net/ieee802154_radio.h (hunk 8 above)
Comment (jfischer-phytec-iot) [reply include/net/ieee802154_radio.h]: I thought rather in the direction to verify which channel is valid.
(It would be more correct to have:
IEEE802154_HW_UWB_BG0 and IEEE802154_HW_UWB_BG1)
So for UWB_BG0 the valid channels are: 0
for UWB_BG1 the valid channels are: 1 .. 15

Comment (tbursztyka) [reply include/net/ieee802154_radio.h]: No need for this, it's addressed in further patches and again, there are so many bands varieties that the L2 won't keep any knowledge of these, it would just eat up ROM (13 carrier bands, on which are applied from 1 to sometime up to 7 different modulations, all of these with different channel limits... it's a mess!). Instead, the drivers will have to provide the supported number of channels, according to the RF settings it currently has installed.

File: subsys/net/ip/l2/ieee802154/ieee802154_radio_utils.h [hunk 9]
```diff
@@ -24,10 +24,15 @@ static inline bool prepare_for_ack(struct ieee802154_context *ctx,
...
+static inline int wait_for_ack(struct net_if *iface,
 			       bool ack_required)
 {
-	if (!ack_required) {
+	const struct ieee802154_radio_api *radio = iface->dev->driver_api;
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+
+	if (!ack_required ||
+	    (radio->get_capabilities(iface->dev) & IEEE802154_HW_TX_RX_ACK)) {
```
Comment (jfischer-phytec-iot) [inline subsys/net/ip/l2/ieee802154/ieee802154_radio_utils.h]: Maybe we need a hint in the API docu, that if the radio supports  IEEE802154_HW_TX_RX_ACK then it should be switched on by default.

**Review by jfischer-no — APPROVED** [review]

**Comment by tbursztyka** [comment]
@wbober if you have time to review this.

**Comment by wbober** [comment]
@hubertmis @KaSroka Could you please take a look a this?

**Comment by tbursztyka** [comment]
@wbober : you don't maintain nrf5 15.4 driver anymore? (I added you because of it)

**Comment by wbober** [comment]
@tbursztyka No. @KaSroka is now responsible for that and @hubermis is the author of the SDK 154 driver.

**Comment by tbursztyka** [comment]
> @tbursztyka No. @KaSroka is now responsible for that and @hubermis is the author of the SDK 154 driver.

Ok!

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 10]
```diff
@@ -137,6 +140,14 @@ static void nrf5_rx_thread(void *arg1, void *arg2, void *arg3)
 
 /* Radio device API */
 
+static enum ieee802154_hw_caps nrf5_get_capabilities(struct device *dev)
+{
+	return IEEE802154_HW_FCS |
+		IEEE802154_HW_2_4_GHZ |
+		IEEE802154_HW_FILTER;
+}
```
Comment (hubertmis) [inline drivers/ieee802154/ieee802154_nrf5.c]: nRF 802.15.4 driver supports automatic ACKs.
It also support promiscuous mode (I'm not sure if promiscuous was tested in version copied to Zephyr repo - it is pretty old but @KaSroka should update it soon).

Comment (tbursztyka) [reply drivers/ieee802154/ieee802154_nrf5.c]: For ack, see last patch of this commit. @KaSroka patch then will need rework on top of that PR.

Promiscuous mode is supported by most chip, it will "just" need  to be implemented on driver side: thus why I did not made it exposed by any yet. (And there is no net_mgmt hooks yet to enter that mode anyway)

**Review by hubertmis — APPROVED** [review]

**Comment by tbursztyka** [comment]
@jukkar can you recheck?

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * @brief IEEE 802.15.4 Management
+ */
+
+#ifndef __IEEE802154_UTILS_H__
+#define __IEEE802154_UTILS_H__
+
+#include <net/ieee802154_radio.h>
+
+static inline void ieee802154_filter_ieee_addr(struct net_if *iface,
+					       u8_t *ieee_addr)
```
Comment (MaureenHelm) [inline subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Why are these functions inlined?

File: subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c [hunk 12]
```diff
@@ -39,7 +39,7 @@ static inline int aloha_tx_fragment(struct net_if *iface,
 			continue;
 		}
 
-		ret = wait_for_ack(ctx, ack_required);
+		ret = wait_for_ack(iface, ack_required);
```
Comment (jfischer-phytec-iot) [inline subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: I started to play with it and noticed it is not enough. The driver must know whether the hardware should wait for an ACK or not. Not every packet may claim an ACK. Since it is not desired that the driver parses the FCS field, the TX function need to be changed like 
`int (*tx)(struct device*dev, struct net_pkt *pkt, struct net_buf *frag, bool ack_required); `

**Review by jfischer-no — CHANGES_REQUESTED** [review]
IEEE802154_HW_TX_RX_ACK needs further changes on ieee802154_radio_api

File: subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c (hunk 12 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: indeed, your mrc20a needs to know that it if I remember well.

That's where chip manufacturers are not consistent: they handle AR flag properly on rx side to quickly respond with an ack frame, which mean their hardware (or inner firmware) is clever enough to parse the FCF to do so. But on tx? nope, letting the driver to do all the heavy lifting.

I think I prefer that drivers do a fcf parse by themselves. I'll add a quick utility function for this (after all it's just a matter of checking one bit in the very first 2 bytes of the ll reserve).

File: include/net/net_pkt.h [hunk 13]
```diff
@@ -122,7 +122,8 @@ struct net_pkt {
 #endif /* CONFIG_NET_IPV6 */
 
 #if defined(CONFIG_NET_L2_IEEE802154)
-	u8_t ieee802154_rssi;
+	u8_t ieee802154_rssi; /* Received Signal Strength Indication */
+	u8_t ieee802154_lqi;  /* Ling Quality Indicator */
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/net/net_pkt.h]: indeed

File: subsys/net/ip/l2/ieee802154/ieee802154_utils.h (hunk 11 above)
Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_utils.h]: Most of them are used in one unique place (only ieee_addr is used twice actually). For the other radio function (rx/tx/etc...), they are going to be one liners. I don't think it's worth having them exposed another way.

File: subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c (hunk 12 above)
Comment (jfischer-phytec-iot) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: There is ieee802154_ack_required(...) in ieee802154_frame.h that may be used for it, but it is not exposed to public.

Comment (tbursztyka) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: Yes but it would mean exposing also the frame structures and all, and that I don't want to.

I'll make a simpler one.

Comment (pfalcon) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: @tbursztyka : @dleach02 raised the same issue in a separate discussion: that a driver may need to look into a packet, and those are defined in subsystem/net/ip/l2/ieee802154/ , instead of include/ . Can you elaborate what's the problem with moving those defines to public includes? We have L3 headers defined publicly, what's the problem with doing the same for standardized L2 technologies?

@dleach02 , I'd let you continue from here, to not serve as a "broken phone" proxy ;-).

Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: I also ran into the "driver needs to know about ACK request" and was going to just look at the AR bit in the FC. But this brings up another problem with our include files because the information on 802.15.4 header structure are in header files in the subsystem and not in the /include area (Linux has it up there). But I also think we may want to consider a transmit API that includes control information on the transmit where one item in this control information is ACK offload requirement.

Comment (dleach02) [reply subsys/net/ip/l2/ieee802154/ieee802154_radio_aloha.c]: I don't think it is bad to expose IEEE802.15.4 frame structures in a more global include that is accessible by the interface drivers. It isn't internal stuff that would be proprietary to the Zephyr 15.4 stack. 

But I realized that I didn't need it yet since the current state of the system is that there isn't ACK offload (except for what is being done here).

I do think there should be a larger discussion about the PLME interface being exposed by the PHY drivers and what future features we want/need to support. The NXP HW has a number of interesting features that we may want to expose through the interface depending on how we wanted the Zephyr stack to handle things (note that I'm new to the NXP HW as well even though I'm an NXP person). For example, when we get around to supporting modes where we want to schedule RX/TX times we would likely want to enable/disable the RX based on this information to conserve power. From what I can see now, the 15.4 stack expects the RX to be enabled all the time, which is likely good enough for now but as we optimize things in the future we may want to consider how to abstract something like this and when to leverage the underlying HW to do some of this.

File: include/net/ieee802154_radio.h [hunk 14]
```diff
@@ -101,6 +122,16 @@ extern enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
...
 void ieee802154_init(struct net_if *iface);
 
+/**
+ * @brief Check if AR flag is set on the frame inside given net_pkt
+ *
+ * @param pkt A valid pointer on a net_pkt structure, must not be NULL.
+ *
+ * @return True if AR flag is set, False otherwise
+ */
+bool ieee802154_is_ar_flag_set(struct net_pkt *pkt);
```
Comment (dleach02) [inline include/net/ieee802154_radio.h]: Not clear to me what the compiler does in this case. It seems that if we are incuring a function call overhead for each time we check this bit that it would be better to just make this an inline/macro. Although I do understand that would require exposing the details of the 802.15.4 protocol frame structure at the /include level... but I'm for doing that anyway.

**Review by jfischer-no — APPROVED** [review]
rebased  #1135 on it and tested, LGTM

**Comment by dleach02** [comment]
What is holding up mergint this PR?

**Review by jukkar — APPROVED** [review]

**Comment by jukkar** [comment]
> What is holding up mergint this PR?

Nothing :) Merging it now.
